File: 8_grafos_arboles/Algoritmo de Backtrack/combination_of_sums.js
function combinationSum(candidates, target) {
    function dfs(remaining, stack, result) {
        if (!remaining) {
            result.push(stack);
            return;
        }
        for (const number of candidates) {
            if (number > remaining) {
                return;
            }
            if (!stack.length || stack[stack.length - 1] <= number) {
                dfs(remaining - number, stack.concat([number]), result);
            }
        }
    }
    const result = [];
    candidates.sort();
    dfs(target, [], result);
    return result;
}

/* This is an example for calling this function: */
const exercise = combinationSum([2, 3, 6, 7], 7);
console.log(exercise); //[[2, 2, 3], [7]]

File: 8_grafos_arboles/Algoritmo de Backtrack/combination_of_sums.py
# Algoritmo de Backtrack para encontrar todas las combinaciones de sumas de un conjunto de números
def combinationSum(candidatos: list[int], target: int) -> list[list[int]]:
    def dfs(sobrante, pila, resultado):
        if not sobrante:
            resultado.append(pila)
            return
        for numero in candidatos:
            if numero > sobrante:
                return
            if not pila or pila[-1] <= numero:
                dfs(sobrante - numero, pila + [numero], resultado)
    resultado = []
    candidatos.sort()
    dfs(target, [], resultado)
    return resultado

""" This is an example for calling this function:"""
exercise = combinationSum([2, 3, 6, 7], 7) 
print(exercise) #[[2, 2, 3], [7]]
